iob_soc.py

```diff
@@ -70,10 +69,10 @@ def _init_attributes(cls):
             )
 
         # warning: do not initialize lists as class variables, they will be shared between all subclasses
-        cls.peripherals = []
-        cls.peripheral_portmap = []
+        cls.init_attr("peripherals", [])
```

Comment:
I don't get this. If this attribute exists why do I have to initialize it?

Reply:
Because the attribute may not exist.
It only exits if there is a subclass of IOb-SoC that defines it before. Otherwise, it will not exist in IOb-SoC.

Reply:
Note that the `peripherals` list is not an attribute of iob_module. So it must be defined in either the iob-soc class or one of its subclasses.
If the iob-soc does not have subclasses, then it must be defined in the iob-soc class. That is why line 72 is there, to ensure that it is defined.

Reply:
There is no SoC without at least one peripheral 

why not 

cls.peripherals = [uart]


?

Reply:
why         cls.init_attr("peripherals", []) ?

        cls.peripherals = [] seems better

Reply:
Because if we do have a subclass, for example opencryptolinux, that subclass may want to initialize its own peripherals list first and then call the `super()._init_attributes` to add the iob-soc peripherals.
So, if the iob-soc has this line `cls.peripherals = []`, then it will override any `cls.peripherals` list that may have been defined previously (by the subclasses).
Example:
```
class iob_soc_opencryptolinux(iob_soc):
    ...
    @classmethod
    def _init_attributes(cls):
        ...
        cls.peripherals = [iob_spi_master]  # Add iob_spi_master peripheral
        super()._init_attributes(cls)  # Call _init_attributes of iob-soc
    ...
```
```
class iob_soc(iob_module):
    ...
    @classmethod
    def _init_attributes(cls):
        ...
        cls.peripherals = []  # Will override any `cls.peripherals` defined previously (like the one from opencryptolinux)
```

Reply:
> There is no SoC without at least one peripheral
> 
> why not
> 
> cls.peripherals = [uart]
> 
> ?

For a similar reason to my previous comment.
If we have a subclass of iob-soc (like opencryptolinux) that has previously defined a list of peripherals, this iob-soc line would override that list.

The current line that appends the UART to the peripherals list is this one: https://github.com/IObundle/iob-soc/pull/634/files#diff-551ef7ec958375dc7823b60fea7a6b8ff010a80b79c53afdd937d66369e41807R218
This line does not override the existing peripherals list, it only appends a UART instance at the end of it.